### ISSUE TITLE
sklearn has deprecated LinearModel Mixins

### ISSUE THREAD

**Issue by peterychang** [VowpalWabbit/vowpal_wabbit#2240]
Related to https://github.com/VowpalWabbit/vowpal_wabbit/issues/2183
Sklearn has deprecated a lot of interfaces into the library in sklearn 0.22. Most of the issues have been addressed in https://github.com/VowpalWabbit/vowpal_wabbit/pull/2208, but there are 2 major objects we'll lose access to with no good replacements.

```
from sklearn.linear_model.base import LinearClassifierMixin, SparseCoefMixin
```
The two Mixin classes from LinearModel have been deprecated as of sklearn 0.22, and will be removed entirely by sklearn 0.24. We'll need to figure out exactly what we lose and how to replace it.

**Comment by kumpera** [comment]
Can you add some sklearn references here from sklearn?
I could not find it.

**Comment by kumpera** [comment]
Here's the PR: https://github.com/scikit-learn/scikit-learn/pull/15324

I wonder what do we lose if we switch to ClassifierMixin/RegressorMixin.

**Comment by peterychang** [comment]
Here's an explanation of the deprecations (though it doesn't say exactly when the private API will be closed off)
https://scikit-learn.org/stable/whats_new/v0.22.html#clear-definition-of-the-public-api

The deprecation warning in sklearn 0.22 mention we'll lose them in 0.24. Their public API is documented here
https://scikit-learn.org/dev/modules/classes.html

**Comment by thealphadollar** [comment]
Hey @peterychang 

What do you think about writing the above classes by ourselves using the Public API instead of importing them from the new version? This will also ensure that in future we can be sure that we don't end up having broken code often.

I believe it should be fairly easy as it'll only require us to write one or two methods by ourselves and use the parent of the `LinearClassifierMixin` to derive our own class.

A similar approach is possible for `SparseCoefMixin`.

**Comment by thealphadollar** [comment]
@jackgerrits @peterychang Any feedback on my comment?

**Comment by thealphadollar** [comment]
I'm starting work on this.

**Comment by peterychang** [comment]
I'm wondering if it would be possible to adapt [sklearn.svm.LinearSVM](https://github.com/scikit-learn/scikit-learn/blob/438bb3a438a42e23e3111c58575cf16841bb694d/sklearn/svm/_classes.py#L12) for this (This object is in the public API).

LinearSVM's inheritance diagram looks similar to the one we use
`class LinearSVC(BaseEstimator, LinearClassifierMixin, SparseCoefMixin)`

vs what we have in sklearn_vw
```
class VW(BaseEstimator)
class ThresholdingLinearClassifierMixin(LinearClassifierMixin)
class VWClassifier(SparseCoefMixin, ThresholdingLinearClassifierMixin, VW)
````

I'm not particularly familiar with python's object inheritance model though, so I'm not sure if we'd be able to modify the behavior in LinearSVM to run what we need.

**Comment by thealphadollar** [comment]
I'll look into this @peterychang . I've quite some experience with inheritance and I'll update the PR accordingly.

**Comment by jackgerrits** [comment]
It looks like this was resolved in #2368
